pkg/sloop/ingress/kubewatcher_test.go

```diff
@@ -0,0 +1,237 @@
+/*
+ * Copyright (c) 2021, salesforce.com, inc.
+ * All rights reserved.
+ * SPDX-License-Identifier: BSD-3-Clause
+ * For full license text, see LICENSE.txt file in the repo root or https://opensource.org/licenses/BSD-3-Clause
+ */
+
+package ingress
+
+import (
+	"encoding/json"
+	"fmt"
+	"sync"
+	"sync/atomic"
+	"testing"
+	"time"
+
+	"github.com/salesforce/sloop/pkg/sloop/store/typed"
+	"github.com/stretchr/testify/assert"
+	corev1 "k8s.io/api/core/v1"
+	apiextensionsv1 "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1"
+	"k8s.io/apiextensions-apiserver/pkg/client/clientset/clientset"
+	clientsetFake "k8s.io/apiextensions-apiserver/pkg/client/clientset/clientset/fake"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/client-go/dynamic/dynamicinformer"
+	dynamicFake "k8s.io/client-go/dynamic/fake"
+	kubernetesFake "k8s.io/client-go/kubernetes/fake"
+	"k8s.io/client-go/rest"
+	k8sTesting "k8s.io/client-go/testing"
+	"k8s.io/client-go/tools/cache"
+)
+
+type dummyData struct {
+	Namespace string
+}
+
+// when fake client tries to list CRDs, return a list with one defined
+func reactionListOfOne(_ k8sTesting.Action) (bool, runtime.Object, error) {
+	versions := []apiextensionsv1.CustomResourceDefinitionVersion{{Name: "v1"}}
+	name := apiextensionsv1.CustomResourceDefinitionNames{Plural: "things", Kind: "k"}
+	spec := apiextensionsv1.CustomResourceDefinitionSpec{Group: "g", Versions: versions, Names: name}
+	crd := apiextensionsv1.CustomResourceDefinition{Spec: spec}
+	list := apiextensionsv1.CustomResourceDefinitionList{Items: []apiextensionsv1.CustomResourceDefinition{crd}}
+	return true, &list, nil
+}
+
+// when fake client tries to list CRDs, return an error
+func reactionError(_ k8sTesting.Action) (bool, runtime.Object, error) {
+	return true, nil, fmt.Errorf("failed")
+}
+
+// newTestCrdClient - provides a function pointer to create a fake clientset
+//   takes: k8sTesting.Action function pointer - adds the reaction to the fake clientset
+//   returns a function pointer
+//      takes: restConfig
+//      returns: clientset.Interface & error (always nil)
+func newTestCrdClient(reaction func(_ k8sTesting.Action) (bool, runtime.Object, error)) func(_ *rest.Config) (clientset.Interface, error) {
```

Comment:
i feel this can be reused, do we want to put it somewhere common so other test can call?

Reply:
I prefer not to make something public/shared until there is at least another consumer. Early publishing of potential code clutters the global/shared namespace. The cost of making something shared later is not a significant barrier.